src/Space.ts

```diff
@@ -244,7 +244,16 @@ class Space extends EventEmitter<SpaceEventMap> {
     return new Promise((resolve) => {
       const presence = this.channel.presence;
 
-      const presenceListener = async () => {
+      const presenceListener = async (presenceMessage: Types.PresenceMessage) => {
+        if (
+          !(
+            presenceMessage.clientId == this.client.auth.clientId &&
+            presenceMessage.connectionId == this.client.connection.id
+          )
+        ) {
+          return;
+        }
```

Comment:
Let's add test coverage for this

Reply:
I agree it would be good to test it, but I don't have any good ideas on how. We want to test that `space.enter()` doesn’t return as a result of receiving a presence message whose client ID or connection ID doesn't match ours, right? I'm not sure of a good way of testing "doesn’t return" without introducing a timeout, something like this:

```typescript
    it<SpaceTestContext>('doesn’t complete as a result of a presence message from a client ID that is not ours', async ({
      space,
      presence,
    }) => {
      let hasEmittedPresenceMessageWithCorrectClientId = false;
      vi.spyOn(presence, 'subscribe').mockImplementation(
        async (_, listener?: (presenceMessage: Types.PresenceMessage) => void) => {
          listener!(
            createPresenceMessage('enter', {
              clientId: 'OTHER_MOCK_CLIENT_ID',
              connectionId: '1',
            }),
          );
          setTimeout(() => {
            hasEmittedPresenceMessageWithCorrectClientId = true;
            listener!(
              createPresenceMessage('enter', {
                clientId: 'MOCK_CLIENT_ID',
                connectionId: '1',
              }),
            );
          }, 500);
        },
      );

      await space.enter();
      expect(hasEmittedPresenceMessageWithCorrectClientId).to.be.true;
    });
```

How do you feel about that?

Reply:
Could you `spyOn` on `subscribe` that it gets called twice but `unsubscribe` just once & expect that the promise only resolves after the second message, not the first? Do you need a timeout here, or could maybe just call them one after the other?

Reply:
ok, had a stab at it - bit different then I described, but how about:
```ts
it<SpaceTestContext>('doesn’t complete as a result of a presence message from a client ID that is not ours', async ({
  space, 
  presence,
}) => {     
  const unsubscribeSpy = vi.spyOn(presence, 'unsubscribe');

  vi.spyOn(presence, 'subscribe').mockImplementation(
    async (_, listener?: (presenceMessage: Types.PresenceMessage) => void) => {
      listener!(
        createPresenceMessage('enter', {
          clientId: 'OTHER_MOCK_CLIENT_ID', // a clientId that does not match our setup mock
          connectionId: '1',
        }),
      );
    },
  );

  Promise.resolve(space.enter())
  expect(unsubscribeSpy).not.toHaveBeenCalled()
}, 1000);


it<SpaceTestContext>('completes as a result of a presence message from a client ID that is ours', async ({
  space,
  presence,
}) => {
  const unsubscribeSpy = vi.spyOn(presence, 'unsubscribe');

  vi.spyOn(presence, 'subscribe').mockImplementation(
    async (_, listener?: (presenceMessage: Types.PresenceMessage) => void) => {
      listener!(
        createPresenceMessage('enter', {
          clientId: 'MOCK_CLIENT_ID', // a clientId that matches our setup mock
          connectionId: '1',
        }),
      );
    },
  );

  Promise.resolve(space.enter())
  expect(unsubscribeSpy).toHaveBeenCalled()
});
```

Reply:
I don’t understand what purpose `Promise.resolve` is serving here, would you mind explaining?

```typescript
  Promise.resolve(space.enter())
```

Am I right in thinking that this is the key idea here — to check whether `presence.unsubscribe` has been called instead of checking whether `space.enter()` has returned?

```typescript
  expect(unsubscribeSpy).not.toHaveBeenCalled()
```

What’s the purpose of this timeout? As far as I can tell, the rest of the test is all synchronous.

```typescript
}, 1000);
```

Reply:
Sorry, the timeout should be removed, was experimenting with how we can test `enter` not resolving.

`Promise.resolve` is a bit of trick - It doesn't do anything except pass the promise returned from `enter`, but it allows us to call the function without an `await`, which would cause the test to wait. This could also be accomplished with an IIFE, albeit it would be more verbose.

>to check whether presence.unsubscribe has been called instead of checking whether space.enter() has returned?
yes, that's the key idea - still not ideal, but avoids any complexity arising from timers, for example (and is fairly concise).

Reply:
> but it allows us to call the function without an `await`

Can't we just do that anyway? An `async` function is just really a function that returns a `Promise`, I thought. Why would we need to `await` its return value in order to be able to call it?

> yes, that's the key idea

Cool, I'll write something based around this.

Reply:
>Can't we just do that anyway? An async function is just really a function that returns a Promise, I thought. Why would we need to await its return value in order to be able to call it?

you can, but then you need to add an exception for typescript (or eslint? can't remember). I don't really mind what we do here.

Reply:
> you can, but then you need to add an exception for typescript (or eslint? can't remember)

Ah, OK — well, I've removed the `Promise.resolve` and nothing seems to be complaining, so I'll keep it like that 😄

Reply:
OK, I've updated the tests in 627b481a9af8454541498650e113314005b4724a, please could you take a look?